File: src/main/java/org/blitzortung/android/alarm/AlertHandler.java
package org.blitzortung.android.alarm;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.location.Location;
import android.media.AudioManager;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Vibrator;
import android.util.Log;
import org.blitzortung.android.AlertResultEvent;
import org.blitzortung.android.alarm.factory.AlarmObjectFactory;
import org.blitzortung.android.alarm.handler.AlarmStatusHandler;
import org.blitzortung.android.alarm.object.AlarmSector;
import org.blitzortung.android.alarm.object.AlarmStatus;
import org.blitzortung.android.app.Main;
import org.blitzortung.android.app.R;
import org.blitzortung.android.location.LocationHandler;
import org.blitzortung.android.app.controller.NotificationHandler;
import org.blitzortung.android.app.view.PreferenceKey;
import org.blitzortung.android.data.beans.Stroke;
import org.blitzortung.android.protocol.Event;
import org.blitzortung.android.protocol.Listener;
import org.blitzortung.android.util.MeasurementSystem;

import java.util.Collection;

public class AlertHandler implements OnSharedPreferenceChangeListener, Listener {

    private final Vibrator vibrator;
    private final NotificationHandler notificationHandler;
    private Context context;
    private Collection<? extends Stroke> lastStrokes;
    private int vibrationSignalDuration;
    private Uri alarmSoundNotificationSignal;

    private final AlarmParameters alarmParameters;

    private Location location;

    private boolean alertEnabled;

    private boolean alarmValid;

    private Listener alertListener;

    private final AlarmStatus alarmStatus;

    private final AlarmStatusHandler alarmStatusHandler;

    private final LocationHandler locationHandler;

    private float notificationDistanceLimit;

    private long notificationLastTimestamp;

    private float signalingDistanceLimit;

    private long signalingLastTimestamp;

    public AlertHandler(LocationHandler locationHandler, SharedPreferences preferences, Context context, Vibrator vibrator, NotificationHandler notificationHandler, AlarmObjectFactory alarmObjectFactory, AlarmParameters alarmParameters) {
        this.locationHandler = locationHandler;
        this.context = context;
        this.vibrator = vibrator;
        this.notificationHandler = notificationHandler;
        this.alarmStatus = alarmObjectFactory.createAlarmStatus(alarmParameters);
        this.alarmStatusHandler = alarmObjectFactory.createAlarmStatusHandler(alarmParameters);
        this.alarmParameters = alarmParameters;


        preferences.registerOnSharedPreferenceChangeListener(this);
        onSharedPreferenceChanged(preferences, PreferenceKey.ALERT_ENABLED);
        onSharedPreferenceChanged(preferences, PreferenceKey.MEASUREMENT_UNIT);
        onSharedPreferenceChanged(preferences, PreferenceKey.ALERT_NOTIFICATION_DISTANCE_LIMIT);
        onSharedPreferenceChanged(preferences, PreferenceKey.ALERT_SIGNALING_DISTANCE_LIMIT);
        onSharedPreferenceChanged(preferences, PreferenceKey.ALERT_VIBRATION_SIGNAL);
        onSharedPreferenceChanged(preferences, PreferenceKey.ALERT_SOUND_SIGNAL);

        alarmValid = false;
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String keyString) {
        onSharedPreferenceChanged(sharedPreferences, PreferenceKey.fromString(keyString));
    }

    private void onSharedPreferenceChanged(SharedPreferences sharedPreferences, PreferenceKey key) {
        switch (key) {
            case ALERT_ENABLED:
                alertEnabled = sharedPreferences.getBoolean(key.toString(), false);

                updateLocationHandler();
                break;

            case MEASUREMENT_UNIT:
                String measurementSystemName = sharedPreferences.getString(key.toString(), MeasurementSystem.METRIC.toString());

                alarmParameters.setMeasurementSystem(MeasurementSystem.valueOf(measurementSystemName));
                break;

            case ALERT_NOTIFICATION_DISTANCE_LIMIT:
                notificationDistanceLimit = Float.parseFloat(sharedPreferences.getString(key.toString(), "50"));
                break;

            case ALERT_SIGNALING_DISTANCE_LIMIT:
                signalingDistanceLimit = Float.parseFloat(sharedPreferences.getString(key.toString(), "25"));
                break;

            case ALERT_VIBRATION_SIGNAL:
                vibrationSignalDuration = sharedPreferences.getInt(key.toString(), 3) * 10;
                break;

            case ALERT_SOUND_SIGNAL:
                final String signalUri = sharedPreferences.getString(key.toString(), "");
                alarmSoundNotificationSignal = !signalUri.isEmpty() ? Uri.parse(signalUri) : null;
                break;
        }
    }

    private void updateLocationHandler() {
        if (alertEnabled && alertListener != null) {
            locationHandler.requestUpdates(this);
        } else {
            locationHandler.removeUpdates(this);
            location = null;
            broadcastClear();
        }
    }

    @Override
    public void onEvent(Event event) {
        this.location = location;

        checkStrokes(lastStrokes);
    }

    public boolean isAlertEnabled() {
        return alertEnabled;
    }

    public void checkStrokes(Collection<? extends Stroke> strokes) {
        boolean currentAlarmIsValid = isAlertEnabled() && location != null && strokes != null;
        lastStrokes = strokes;

        if (currentAlarmIsValid) {
            alarmValid = true;
            alarmStatusHandler.checkStrokes(alarmStatus, strokes, location);
            processResult(getAlarmResult());
        } else {
            invalidateAlert();
        }
    }

    public AlarmResult getAlarmResult() {
        return alarmValid ? alarmStatusHandler.getCurrentActivity(alarmStatus) : null;
    }

    public String getTextMessage(float notificationDistanceLimit) {
        return alarmStatusHandler.getTextMessage(alarmStatus, notificationDistanceLimit);
    }

    public void setAlertListener(Listener alertListener) {
        this.alertListener = alertListener;
        updateLocationHandler();
    }

    public void unsetAlertListener() {
        alertListener = null;
        updateLocationHandler();
    }

    public Collection<AlarmSector> getAlarmSectors() {
        return alarmStatus.getSectors();
    }

    public AlarmParameters getAlarmParameters() {
        return alarmParameters;
    }

    public AlarmStatus getAlarmStatus() {
        return alarmValid ? alarmStatus : null;
    }

    public float getMaxDistance() {
        final float[] ranges = alarmParameters.getRangeSteps();
        return ranges[ranges.length - 1];
    }

    public void invalidateAlert() {
        boolean previousAlarmValidState = alarmValid;
        alarmValid = false;

        if (previousAlarmValidState) {
            alarmStatus.clearResults();
            broadcastClear();
        }
    }

    private void broadcastClear() {
        if (alertListener != null) {
            alertListener.onEvent(new AlertCancelEvent());
        }
    }

    private void broadcastResult(AlarmResult alarmResult) {
        if (alertListener != null) {
            alertListener.onEvent(new AlertResultEvent(alarmStatus, alarmResult));
        }
    }

    private void processResult(AlarmResult alarmResult) {
        if (alarmResult != null) {

            alarmParameters.updateSectorLabels(context);

            if (alarmResult.getClosestStrokeDistance() <= signalingDistanceLimit) {
                long signalingLatestTimestamp = alarmStatusHandler.getLatestTimstampWithin(signalingDistanceLimit, alarmStatus);
                if (signalingLatestTimestamp > signalingLastTimestamp) {
                    Log.v(Main.LOG_TAG, "AlertHandler.processResult() perform alarm");
                    vibrateIfEnabled();
                    playSoundIfEnabled();
                    signalingLastTimestamp = signalingLatestTimestamp;
                } else {
                    Log.d(Main.LOG_TAG, String.format("old signaling event: %d vs %d", signalingLatestTimestamp, signalingLastTimestamp));
                }
            }

            if (alarmResult.getClosestStrokeDistance() <= notificationDistanceLimit) {
                long notificationLatestTimestamp = alarmStatusHandler.getLatestTimstampWithin(notificationDistanceLimit, alarmStatus);
                if (notificationLatestTimestamp > notificationLastTimestamp) {
                    Log.v(Main.LOG_TAG, "AlertHandler.processResult() perform notification");
                    notificationHandler.sendNotification(context.getResources().getString(R.string.activity) + ": " + getTextMessage(notificationDistanceLimit));
                    notificationLastTimestamp = notificationLatestTimestamp;
                } else {
                    Log.d(Main.LOG_TAG, String.format("AlertHandler.processResult() previous signaling event: %d vs %d", notificationLatestTimestamp, signalingLastTimestamp));
                }
            } else {
                notificationHandler.clearNotification();
            }
        } else {
            notificationHandler.clearNotification();
        }

        Log.v(Main.LOG_TAG, String.format("AlertHandler.processResult() broadcast result %s", alarmResult));

        broadcastResult(alarmResult);
    }

    private void vibrateIfEnabled() {
        vibrator.vibrate(vibrationSignalDuration);
    }

    private void playSoundIfEnabled() {
        if (alarmSoundNotificationSignal != null) {
            Ringtone r = RingtoneManager.getRingtone(context, alarmSoundNotificationSignal);
            if (r != null) {
                if (!r.isPlaying()) {
                    r.setStreamType(AudioManager.STREAM_NOTIFICATION);
                    r.play();
                }
                Log.v(Main.LOG_TAG, "playing " + r.getTitle(context));
            }
        }
    }

    public Location getCurrentLocation() {
        return location;
    }

    public AlertEvent getAlertEvent() {
        return alarmValid ? new AlertResultEvent(alarmStatus, getAlarmResult()) : new AlertCancelEvent();
    }
}

File: app/src/main/java/org/blitzortung/android/alert/handler/AlertHandler.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.alert.handler

import android.content.Context
import android.content.SharedPreferences
import android.location.Location
import android.util.Log
import javax.inject.Inject
import javax.inject.Singleton
import org.blitzortung.android.alert.LocalActivity
import org.blitzortung.android.alert.AlertParameters
import org.blitzortung.android.alert.Warning
import org.blitzortung.android.alert.NoData
import org.blitzortung.android.alert.NoLocation
import org.blitzortung.android.app.Main
import org.blitzortung.android.app.R
import org.blitzortung.android.app.controller.NotificationHandler
import org.blitzortung.android.app.view.OnSharedPreferenceChangeListener
import org.blitzortung.android.app.view.PreferenceKey
import org.blitzortung.android.app.view.get
import org.blitzortung.android.data.beans.GridParameters
import org.blitzortung.android.data.beans.Strike
import org.blitzortung.android.data.provider.result.DataEvent
import org.blitzortung.android.data.provider.result.DataReceived
import org.blitzortung.android.location.LocationEvent
import org.blitzortung.android.location.LocationHandler
import org.blitzortung.android.location.LocationUpdate
import org.blitzortung.android.protocol.ConsumerContainer
import org.blitzortung.android.util.MeasurementSystem

@Singleton
class AlertHandler
    @Inject
    constructor(
        private val locationHandler: LocationHandler,
        preferences: SharedPreferences,
        private val context: Context,
        private val notificationHandler: NotificationHandler,
        private val alertDataHandler: AlertDataHandler,
        private val alertSignal: AlertSignal,
    ) : OnSharedPreferenceChangeListener {
        var alertParameters: AlertParameters
            private set

        private val alertConsumerContainer: ConsumerContainer<Warning> =
            object : ConsumerContainer<Warning>(NoData) {
                override fun addedFirstConsumer() {
                    Log.d(Main.LOG_TAG, "AlertHandler: added first alert consumer")
                }

                override fun removedLastConsumer() {
                    Log.d(Main.LOG_TAG, "AlertHandler: removed last alert consumer")
                }
            }

        private var lastStrikes: Strikes? = null

        private var alertEnabled: Boolean = false

        private var notificationDistanceLimit: Float = 0.0f

        private var notificationLastTimestamp: Long = 0

        private var signalingDistanceLimit: Float = 0.0f

        private var signalingThresholdTime: Long = 0

        private var signalingLastTimestamp: Long = 0

        private val locationEventConsumer: (LocationEvent) -> Unit = { event ->
            Log.v(Main.LOG_TAG, "AlertHandler.locationEventConsumer ${event}")

            if (event is LocationUpdate) {
                checkStrikes(lastStrikes, event.location)
            }
        }

        init {
            locationHandler.requestUpdates(locationEventConsumer)
        }

        val dataEventConsumer: (DataEvent) -> Unit = { event ->
            if (event is DataReceived) {
                if (!event.flags.ignoreForAlerting) {
                    Log.v(Main.LOG_TAG, "AlertHandler.dataEventConsumer $event")
                    if (!event.failed && event.containsRealtimeData() && event.strikes != null) {
                        val strikes = Strikes(event.strikes, event.gridParameters)
                        checkStrikes(strikes, locationHandler.location)
                    } else {
                        if (!event.containsRealtimeData()) {
                            lastStrikes = null
                        }
                        broadcastResult(NoData)
                    }
                }
            }
        }

        init {
            Log.d(Main.LOG_TAG, "AlertHandler() create $this")
            val rangeSteps = listOf(10f, 25f, 50f, 100f, 250f, 500f)
            val alarmInterval = 10 * 60 * 1000L
            val sectorLabels = context.resources.getStringArray(R.array.direction_names).toList()
            alertParameters = AlertParameters(alarmInterval, rangeSteps, sectorLabels, MeasurementSystem.METRIC)

            preferences.registerOnSharedPreferenceChangeListener(this)
            onSharedPreferenceChanged(
                preferences,
                PreferenceKey.ALERT_ENABLED,
                PreferenceKey.MEASUREMENT_UNIT,
                PreferenceKey.ALERT_NOTIFICATION_DISTANCE_LIMIT,
                PreferenceKey.ALERT_SIGNALING_DISTANCE_LIMIT,
                PreferenceKey.ALERT_SIGNALING_THRESHOLD_TIME,
                PreferenceKey.ALERT_VIBRATION_SIGNAL,
                PreferenceKey.ALERT_SOUND_SIGNAL,
            )
        }

        override fun onSharedPreferenceChanged(
            sharedPreferences: SharedPreferences,
            key: PreferenceKey,
        ) {
            when (key) {
                PreferenceKey.ALERT_ENABLED -> {
                    alertEnabled = sharedPreferences.get(key, false)
                    Log.v(Main.LOG_TAG, "AlertHandler.onSharedPreferenceChanged() alertEnabled = $alertEnabled")
                }

                PreferenceKey.MEASUREMENT_UNIT -> {
                    val measurementSystemName = sharedPreferences.get(key, MeasurementSystem.METRIC.toString())

                    alertParameters =
                        alertParameters.copy(measurementSystem = MeasurementSystem.valueOf(measurementSystemName))
                    Log.v(
                        Main.LOG_TAG,
                        "AlertHandler.onSharedPreferenceChanged() measurementSystem = ${alertParameters.measurementSystem}",
                    )
                }

                PreferenceKey.ALERT_NOTIFICATION_DISTANCE_LIMIT -> {
                    notificationDistanceLimit = sharedPreferences.get(key, "50").toFloat()
                    Log.v(
                        Main.LOG_TAG,
                        "AlertHandler.onSharedPreferenceChanged() notificationDistanceLimit = $notificationDistanceLimit",
                    )
                }

                PreferenceKey.ALERT_SIGNALING_DISTANCE_LIMIT -> {
                    signalingDistanceLimit = sharedPreferences.get(key, "25").toFloat()
                    Log.v(
                        Main.LOG_TAG,
                        "AlertHandler.onSharedPreferenceChanged() signalingDistanceLimit = $signalingDistanceLimit",
                    )
                }

                PreferenceKey.ALERT_SIGNALING_THRESHOLD_TIME -> {
                    signalingThresholdTime = sharedPreferences.get(key, "25").toLong() * 1000 * 60
                    Log.v(
                        Main.LOG_TAG,
                        "AlertHandler.onSharedPreferenceChanged() signalingThresholdTime = $signalingThresholdTime",
                    )
                }

                else -> {}
            }
        }

        private fun checkStrikes(
            strikes: Strikes?,
            location: Location?,
        ) {
            lastStrikes = strikes

            val alertResult =
                if (location != null && strikes != null) {
                    alertDataHandler.checkStrikes(strikes, location, alertParameters)
                } else {
                    Log.v(
                        Main.LOG_TAG,
                        "AlertHandler.checkStrikes() strikes: ${strikes != null}, location: ${locationHandler.location != null}",
                    )
                    if (location == null) {
                        NoLocation
                    } else {
                        NoData
                    }
                }

            processResult(alertResult)
        }

        val maxDistance: Float
            get() = alertParameters.rangeSteps.last()

        private fun broadcastResult(warning: Warning) {
            Log.d(Main.LOG_TAG, "AlertHandler.broadcastResult($warning)")
            alertConsumerContainer.storeAndBroadcast(warning)
        }

        private fun processResult(warning: Warning) {
            if (alertEnabled && warning is LocalActivity) {
                if (warning.closestStrikeDistance <= signalingDistanceLimit) {
                    alertSignal(warning)
                }
                if (warning.closestStrikeDistance <= notificationDistanceLimit) {
                    alertNotification(warning)
                }
            }
            broadcastResult(warning)
        }

        private fun alertSignal(alertResult: LocalActivity) {
            val signalingLatestTimestamp = alertDataHandler.getLatestTimstampWithin(signalingDistanceLimit, alertResult)
            if (signalingLatestTimestamp > signalingLastTimestamp + signalingThresholdTime) {
                Log.d(Main.LOG_TAG, "AlertHandler.alertSignal() signal ${signalingLatestTimestamp / 1000}")
                alertSignal.emitSignal()
                signalingLastTimestamp = signalingLatestTimestamp
            } else {
                Log.d(
                    Main.LOG_TAG,
                    "AlertHandler.alertSignal() skipped - ${(signalingLatestTimestamp - signalingLastTimestamp) / 1000}, threshold: ${signalingThresholdTime / 1000}",
                )
            }
        }

        private fun alertNotification(alarm: LocalActivity) {
            val notificationLatestTimestamp =
                alertDataHandler.getLatestTimstampWithin(notificationDistanceLimit, alarm)
            if (notificationLatestTimestamp > notificationLastTimestamp) {
                Log.d(
                    Main.LOG_TAG,
                    "AlertHandler.alertNotification() notification ${notificationLatestTimestamp / 1000}",
                )
                notificationHandler.sendNotification(
                    context.resources.getString(R.string.activity) + ": " +
                        alertDataHandler.getTextMessage(
                            alarm,
                            notificationDistanceLimit,
                            context.resources,
                        ),
                )
                notificationLastTimestamp = notificationLatestTimestamp
            } else {
                Log.d(
                    Main.LOG_TAG,
                    "AlertHandler.alertNotification() skipped ${notificationLatestTimestamp - notificationLastTimestamp}",
                )
            }
        }

        fun requestUpdates(alertEventConsumer: (Warning) -> Unit) {
            alertConsumerContainer.addConsumer(alertEventConsumer)
        }

        fun removeUpdates(alertEventConsumer: (Warning) -> Unit) {
            alertConsumerContainer.removeConsumer(alertEventConsumer)
        }

        val alertEvent: Warning
            get() = alertConsumerContainer.currentPayload
    }

data class Strikes(
    val strikes: Collection<Strike>,
    val gridParameters: GridParameters? = null,
)
